File: SampleApp/basefiles/src/main/java/com/rp/basefiles/IBaseAdapterPresenter.java
package com.rp.basefiles;

import android.support.v7.util.DiffUtil;

import com.rp.util.adapter.RAdapterPayloadWatcher;

import java.util.List;

/**
 * Created by rahul on 14/1/18.
 */

public interface IBaseAdapterPresenter<E> {

    void onAttach(IBaseHolderView baseHolderView);
    void onAttachAdapter(BaseAdapter adapter);
    void onAttachDiffCallback(DiffUtil.ItemCallback<E> callback);

    void onAttachPayloadWatcher(RAdapterPayloadWatcher<E> payloadWatcher);

    int getCount();
    void onBind(int position);
    void onBind(int position, Object payloads);

    void update(List<E> list);
    void addNewList(List<E> listNewItems);
    void addItem(E data);
    void addItemAt(int position, E data);
    void removeItem(int position);

    E getFrom(int position);
    List<E> getAll();
}

File: SampleApp/basefiles/src/main/java/com/rp/basefiles/IBaseAdapterPresenter.kt
package com.rp.basefiles

import androidx.recyclerview.widget.DiffUtil
import com.rp.util.adapter.RAdapterPayloadWatcher

interface IBaseAdapterPresenter<E> {

    val count: Int
    val all: List<E>

    fun onAttach(baseHolderView: IBaseHolderView)
    fun onAttachAdapter(adapter: BaseAdapter<*, *>)
    fun onAttachDiffCallback(callback: DiffUtil.ItemCallback<E>)

    fun onAttachPayloadWatcher(payloadWatcher: RAdapterPayloadWatcher<E>)

    fun update(list: List<E>)

    fun addNewList(listNewItems: List<E>)
    fun addItem(data: E)
    fun addItemAt(position: Int, data: E)

    fun removeItem(position: Int)

    fun changeItemAt(position: Int, data: E)

    fun getFrom(position: Int): E?

    fun onBind(position: Int)
    fun onBind(position: Int, payloads: Any)

    fun filter(query : String)
}